waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +647,192 @@
+                          starting-instance-id->start-timestamp] :as tracker-state}]
+                 service-id->launch-tracker']
+      service-id
+      (let [healthy-instances (get service-id->healthy-instances service-id)
+            healthy-instance-id? (->> healthy-instances (map :id) set)
+            known-instances' (->> service-id
+                                  (get service-id->unhealthy-instances)
+                                  (concat healthy-instances))
+            previously-known-instance? (comp known-instance-ids :id)
+            new-instance-ids (->> known-instances'
+                                  (remove previously-known-instance?)
+                                  (sort instance-comparator)
+                                  (mapv :id))
+            known-instance-ids' (->> known-instances' (map :id) set)
+            removed-instance-ids (remove known-instance-ids known-instance-ids')
+            ;; React to upward-scaling
+            instance-counts' (get service-id->instance-counts service-id instance-counts-zero)
+            instances-requested-delta (- (:requested instance-counts')
+                                         (:requested instance-counts))
+            instance-scheduling-start-times'
+            (->> current-time
+                 (repeat instances-requested-delta)
+                 (into instance-scheduling-start-times))
+            ;; Track scheduling instances
+            unmatched-instance-count (- (count new-instance-ids)
+                                        (count instance-scheduling-start-times'))
+            [matched-start-times instance-scheduling-start-times'']
+            (split-at (count new-instance-ids) instance-scheduling-start-times')
+            ;; Track starting instances
+            new-instance-id->start-timestamp
+            (pc/map-from-keys (constantly current-time) new-instance-ids)
+            starting-instance-id->start-timestamp'
+            (->> removed-instance-ids
+                 (reduce dissoc starting-instance-id->start-timestamp)
+                 (merge new-instance-id->start-timestamp))
+            {started-instance-ids :started starting-instances-ids :starting}
+            (group-by #(if (healthy-instance-id? %) :started :starting)
+                      (keys starting-instance-id->start-timestamp'))
+            starting-instance-id->start-timestamp''
+            (select-keys starting-instance-id->start-timestamp' starting-instances-ids)]
+        ;; Warn about any new instances for which we didn't track the scheduling time
+        (when (pos? unmatched-instance-count)
+          (log/warn unmatched-instance-count
+                    "untracked scheduled instances discovered in instance launch metric loop for service"
+                    service-id))
+        ;; Report schedule time for now-known instances
+        (doseq [start-time matched-start-times]
```

Comment:
Nitpick: should we consider not reporting these numbers for newly known instance ids as they will incorrectly contribute a time of almost 0 to this metric?

Reply:
I think it would be much worse to not report these. Having these reported tells us that the services are being scheduled in < 10 seconds, but excluding these 1) makes it look like the service isn't scheduling any new instances, and/or 2) causes p95 timings to look like p50 timings.